readme.md
# hapi-password  


[![Build Status](https://travis-ci.org/firstandthird/hapi-password.svg?branch=master)](https://travis-ci.org/firstandthird/hapi-password)


### A Hapi plugin that provides an easy-to-use password authentication scheme and login page for your routes.  


## 1. Install

```
npm install hapi-password
```

## 2. Use

```js
// register the plugin with your hapi server:
server.register({
  register: require('hapi-password'),
  options: {
    path: '/myCookiePath'
  }
}, () => {
  // now the 'password' auth strategy is available:
  server.auth.strategy('password', 'password', true, {
    // the password to allow:
    password: {
      'my_password': {
        // some optional credentials information:
        name: 'Thing One'
      },
      'another password': {
        // some optional credentials information:
        name: 'Thing Two'
      }
    },
    // you can define a salt:
    salt: 'here is a random salt for encryption',
    // you can customize the name of your cookie:
    cookieName: 'my-cookie',
    // how long should the cookie live?
    ttl: 1000 * 60 * 5,
    // the name of the query parameter that will have the key:
    queryKey: 'token',
    // hapi-password comes equipped with an HTML login template
    // you can customize how it looks or make your own:
    loginForm: {
      name: 'The Login Form',
      description: 'Enter your password to proceed.'
    }
  });
});
```

  Now your routes are password protected!  Invoking a protected route will cause hapi to redirect the browser to the '/login' page. Once logged in, session authentication is provided by the cookie, which is fully configurable and lasts until the ttl timer expires.
  To review, the steps were:
  1. create a hapi server
  2. register the hapi-password plugin with your server
  3. register the strategy with whatever options you want
  4. define your routes

   See the example folder and the unit tests in test/server-test.js for more examples.

## 3. Strategy options
  These are options that you can pass to your call to server.auth.strategy:

- __password__: a password or map of passwords -> user credentials. If only one password is provided, then all users will share that one common password and one common account. Alternatively you can provide an object, where the keys are the passwords and the values are the [hapi user credentials](http://hapijs.com/tutorials/auth#authenticate) object associated with that password.
- __salt__: a [salt](https://en.wikipedia.org/wiki/Salt_(cryptography) used for encrypting passwords.
- __cookieName__: the name to give the authentication cookie
- __ttl__: how long (in milliseconds) before the cookie is unset and the authentication expires
- __queryKey__: the query parameter that specifies the key
- __loginForm__: data to pass to the login.html view

## 4. Registration Options

  Thse are options that you can pass to your call to server.register, when registering the plugin:
- __path__: the path of the authentication cookie (see https://www.nczonline.net/blog/2009/05/05/http-cookies-explained/ for help understanding cookie paths).  Default is '/'.
- __schemeName__: the name hapi will use to identify this authentication scheme, can be whatever you want.  The default is "password".


CHANGELOG.md

3.2.0 / 2017-02-02
==================

  * refactored into separate files
  * renderLogin for both /login and strategy
  * add test
  * update example\server.js
  * chore(package): update dependencies

3.1.0 / 2017-01-12
==================

  * Option to log signin attempts

3.0.0 / 2017-01-10
==================

  * Redirects if already logged in

2.2.0 / 2016-10-21
==================

  * added name to creds

2.1.1 / 2016-10-21
==================

  * fixed ttl

2.1.0 / 2016-10-21
==================

  * updated readme
  * path defaults to /
  * ttl defaults to 30 days
  * name and password use same ttl
  * updated to hapi 15, set isSecure to false

2.0.0 / 2016-07-05
==================

  * allows options.password to be specified as an object

1.1.0 / 2016-06-20
==================

  * allows non-RFC 6265 cookies for 'name'

1.0.0 / 2016-06-15
==================

  * removed query.api/queryKey options
  * updated package and lib/ example/ files
  * travis, readme, eslint added, lab tests added

0.6.0 / 2016-06-12
==================

  * added option for logout endpoint
  * option to ask users name when logging in
  * don't limit example server to localhost
  * updated depedencies
  * added example npm script

0.5.0 / 2015-10-23
==================

  * set cookie on root path

0.4.0 / 2015-09-26
==================

  * pass in text password via query instead of salted
  * must pass in salt
  * allow setting or disabling the query key.  fixing so it actually works
  * remove default password and throw error if nothing passed in

0.3.0 / 2015-05-13
==================

  * fixed spacing if no description
  * be able to set a ttl for the password cookie


0.2.0 / 2015-04-28
==================

  * fix when authenticated


0.1.1 / 2015-04-23
==================

  * tweak to login page design


0.1.0 / 2015-04-23
==================

  * refactor, better default login screen, updated options - see example


0.0.1 / 2014-11-01 
==================

  * Initial Release


test/server-test.js
'use strict';
const Hapi = require('hapi');
const code = require('code');
const lab = exports.lab = require('lab').script();
const hapiPassword = require('../index.js');

let server;
lab.beforeEach((done) => {
  server = new Hapi.Server({
    debug: {
      log: ['hapi-password']
    }
  });
  server.connection({ port: 8080 });
  done();
});

lab.afterEach((done) => {
  server.stop(() => {
    done();
  });
});

lab.test('should redirect if credentials not posted ', (done) => {
  server.register({
    register: hapiPassword,
    options: {}
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/',
      config: {
        handler: (request, reply) => {
          reply(request.auth);
        }
      }
    });
    server.inject({
      url: '/'
    }, (response) => {
      code.expect(response.statusCode).to.equal(302);
      code.expect(response.headers.location).to.equal('/login?next=/');
      done();
    });
  });
});

lab.test('should not redirect /login if credentials not posted ', (done) => {
  server.register({
    register: hapiPassword,
    options: {}
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.inject({
      url: '/login'
    }, (response) => {
      code.expect(response.statusCode).to.equal(200);
      done();
    });
  });
});

lab.test('passes back a security cookie when credentials are posted ', (done) => {
  server.register({
    register: hapiPassword,
    options: {
      cookieName: 'demo-login',
      isSecure: true
    }
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/success',
      config: {
        handler: (request, reply) => {
          return reply('success!');
        }
      }
    });
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'somename',
          password: 'password',
          next: '/success'
        }
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/success');
        code.expect(response.headers['set-cookie']).to.not.equal(undefined);
        code.expect(response.headers['set-cookie'][0]).to.include('demo-login');
        code.expect(response.headers['set-cookie'][0]).to.include('Secure;');
        done();
      });
    });
  });
});

lab.test('allows login when credentials are posted ', (done) => {
  server.register({
    register: hapiPassword,
    options: {
      cookieName: 'demo-login',
      isSecure: true
    }
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/success',
      config: {
        handler: (request, reply) => {
          return reply('success!');
        }
      }
    });
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'somename',
          password: 'password',
          next: '/success'
        }
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/success');
        code.expect(response.headers['set-cookie']).to.not.equal(undefined);
        code.expect(response.headers['set-cookie'][0]).to.include('demo-login');
        const cookieString = `${response.headers['set-cookie'][0].split(';')[0]};`;
        server.inject({
          url: response.headers.location,
          headers: {
            Cookie: cookieString
          }
        }, (getResponse) => {
          code.expect(getResponse.statusCode).to.equal(200);
          code.expect(getResponse.result).to.equal('success!');
          done();
        });
      });
    });
  });
});

lab.test('allows login and logout ', (done) => {
  server.register({
    register: hapiPassword,
    options: {
      cookieName: 'demo-login',
      isSecure: true
    }
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/success',
      config: {
        handler: (request, reply) => {
          return reply('success!');
        }
      }
    });
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'somename',
          password: 'password',
          next: '/success'
        }
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/success');
        code.expect(response.headers['set-cookie']).to.not.equal(undefined);
        code.expect(response.headers['set-cookie'][0]).to.include('demo-login');
        const cookieString = `${response.headers['set-cookie'][0].split(';')[0]};`;
        server.inject({
          url: '/success',
          headers: {
            Cookie: cookieString
          }
        }, (getResponse) => {
          code.expect(getResponse.statusCode).to.equal(200);
          code.expect(getResponse.result).to.equal('success!');
          server.inject({
            url: '/logout',
            headers: {
              Cookie: cookieString
            }
          }, (getResponse2) => {
            code.expect(getResponse2.statusCode).to.equal(302);
            code.expect(getResponse2.headers['set-cookie'][0]).to.include('demo-login=;');
            const cookieString2 = `${getResponse2.headers['set-cookie'][0].split(';')[0]};`;
            server.inject({
              url: '/success',
              headers: {
                Cookie: cookieString2
              }
            }, (loggedOutResponse) => {
              code.expect(loggedOutResponse.statusCode).to.equal(302);
              code.expect(loggedOutResponse.headers.location).to.equal('/login?next=/success');
              done();
            });
          });
        });
      });
    });
  });
});

lab.test('login route redirects if user already logged in ', (done) => {
  server.register({
    register: hapiPassword,
    options: {}
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/success',
      config: {
        auth: 'password',
        handler: (request, reply) => {
          return reply('success!');
        }
      }
    });
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'somename',
          password: 'password',
          next: '/success'
        }
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/success');
        const cookieString = `${response.headers['set-cookie'][0].split(';')[0]};`;
        server.inject({
          url: '/login',
          method: 'GET',
          headers: {
            Cookie: cookieString
          }
        }, (getResponse) => {
          code.expect(getResponse.statusCode).to.equal(302);
          done();
        });
      });
    });
  });
});

lab.test('allows you to specify multiple credentials to match against ', (done) => {
  server.register({
    register: hapiPassword,
    options: {
      password: {
        'a password': {
          // some optional credentials information:
          name: 'Who Is There'
        },
        'another password': {
          // some optional credentials information:
          name: 'Interrupting Cow'
        }
      },
      cookieName: 'demo-login',
    }
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/success',
      config: {
        handler: (request, reply) => {
          return reply('success!');
        }
      }
    });
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'somename',
          password: 'another password',
          next: '/success'
        }
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/success');
        code.expect(response.headers['set-cookie']).to.not.equal(undefined);
        code.expect(response.headers['set-cookie'][0]).to.include('demo-login');
        const cookieString = `${response.headers['set-cookie'][0].split(';')[0]};`;
        server.inject({
          url: response.headers.location,
          headers: {
            Cookie: cookieString
          }
        }, (getResponse) => {
          code.expect(getResponse.statusCode).to.equal(200);
          code.expect(getResponse.result).to.equal('success!');
          done();
        });
      });
    });
  });
});

lab.test('returns the correct credentials for a given password ', (done) => {
  server.register({
    register: hapiPassword,
    options: {
      cookieName: 'demo-login',
      password: {
        'a password': {
          // some optional credentials information:
          name: 'Who Is There',
          role: 'serf'
        },
        'another password': {
          // some optional credentials information:
          name: 'Interrupting Cow',
          role: 'admin'
        }
      }
    }
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/success',
      config: {
        handler: (request, reply) => {
          return reply(request.auth.credentials);
        }
      }
    });
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'somename',
          password: 'another password',
          next: '/success'
        }
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/success');
        code.expect(response.headers['set-cookie']).to.not.equal(undefined);
        code.expect(response.headers['set-cookie'][0]).to.include('demo-login');
        const cookieString = `${response.headers['set-cookie'][0].split(';')[0]};`;
        server.inject({
          url: response.headers.location,
          headers: {
            Cookie: cookieString
          }
        }, (getResponse) => {
          code.expect(getResponse.statusCode).to.equal(200);
          // test will go here:
          code.expect(typeof getResponse.result).to.equal('object');
          code.expect(getResponse.result.name).to.equal('Interrupting Cow');
          code.expect(getResponse.result.role).to.equal('admin');
          done();
        });
      });
    });
  });
});

lab.test('allows login when credentials are posted even if name has a space in it', (done) => {
  server.register({
    register: hapiPassword,
    options: {
      cookieName: 'demo-login',
      queryKey: 'token'
    }
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/success',
      config: {
        handler: (request, reply) => {
          return reply('success!');
        }
      }
    });
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'No One',
          password: 'password',
          next: '/success'
        }
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/success');
        code.expect(response.headers['set-cookie']).to.not.equal(undefined);
        code.expect(response.headers['set-cookie'][0]).to.include('demo-login');
        const cookieString = `${response.headers['set-cookie'][0].split(';')[0]};`;
        server.inject({
          url: response.headers.location,
          headers: {
            Cookie: cookieString
          }
        }, (getResponse) => {
          code.expect(getResponse.statusCode).to.equal(200);
          code.expect(getResponse.result).to.equal('success!');
          done();
        });
      });
    });
  });
});

lab.test('able to pass in isSecure option to cookie setting', (done) => {
  server.register({
    register: hapiPassword,
    options: {
      isSecure: false,
      cookieName: 'demo-login'
    }
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/success',
      config: {
        handler: (request, reply) => {
          return reply('success!');
        }
      }
    });
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'somename',
          password: 'password',
          next: '/success'
        }
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/success');
        code.expect(response.headers['set-cookie']).to.not.equal(undefined);
        code.expect(response.headers['set-cookie'][0]).to.include('demo-login');
        // will not contain the Secure header:
        code.expect(response.headers['set-cookie'][0]).to.not.include('Secure;');
        done();
      });
    });
  });
});

lab.test('path as option, default path is "/"', (done) => {
  server.register({
    register: hapiPassword,
    options: {
      path: '/path1/path2',
      cookieName: 'demo-login'
    }
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/success',
      config: {
        handler: (request, reply) => {
          return reply('success!');
        }
      }
    });
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'somename',
          password: 'password',
          next: '/success'
        }
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/success');
        code.expect(response.headers['set-cookie']).to.not.equal(undefined);
        code.expect(response.headers['set-cookie'][0]).to.include('demo-login');
        code.expect(response.headers['set-cookie'][1]).to.include('Path=/path1/path2');
        done();
      });
    });
  });
});

lab.test('option to log failed passwords', (done) => {
  const results = [];
  const oldLog = console.error;
  console.error = (t1, t2, t3) => {
    results.push(t1);
    results.push(t2);
    results.push(t3);
  };
  server.register({
    register: hapiPassword,
    options: {
      logFailedAttempts: true
    }
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'somename',
          password: 'incorrect',
          next: '/success'
        }
      }, (response) => {
        console.error = oldLog;
        code.expect(response.statusCode).to.equal(302);
        code.expect(results.length).to.equal(3);
        code.expect(results[2]).to.include('Failed login at');
        done();
      });
    });
  });
});

lab.test('can stay logged in with "?token="', (done) => {
  server.register({
    register: hapiPassword,
    options: {}
  }, (err) => {
    if (err) {
      console.log(err);
    }
    server.route({
      method: 'GET',
      path: '/success',
      config: {
        auth: 'password',
        handler: (request, reply) => {
          return reply('success!');
        }
      }
    });
    server.start(() => {
      server.inject({
        url: '/login',
        method: 'POST',
        payload: {
          name: 'somename',
          password: 'password',
          next: '/success'
        }
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/success');
        const cookieString = `${response.headers['set-cookie'][0].split(';')[0].split('=')[1]};`;
        server.inject({
          url: `/login?token=${cookieString}`,
          method: 'GET',
        }, (getResponse) => {
          code.expect(getResponse.statusCode).to.equal(302);
          code.expect(response.headers.location).to.equal('/success');
          done();
        });
      });
    });
  });
});


lib/authScheme.js
'use strict';
const Hoek = require('hoek');
const renderLogin = require('./renderLogin.js');
const lookup = require('./credentials.js');
const _ = require('lodash');

module.exports = (server, pluginOptions, schemeDefaults) => {
  server.auth.scheme(pluginOptions.schemeName, (authServer, options) => {
    options = Hoek.applyToDefaults(schemeDefaults, options);
    lookup.registerCredentials(options.password, options.salt);
    return {
      authenticate: (request, reply) => {
        const cookie = request.state[options.cookieName] ? request.state[options.cookieName] : request.query[options.queryKey];
        const name = request.state[options.cookieNameName];
        if (cookie) {
          const credentials = lookup.getCredentials(cookie, true);
          if (credentials) {
            // return the credentials, minus the encryptedPassword:
            const creds = _.omit(credentials, ['encryptedPassword']);
            if (!creds.name && name) {
              creds.name = name;
            }
            // 'credentials' is either the credentials object or just contains it:
            if (credentials.credentials) {
              return reply.continue(credentials);
            }
            return reply.continue({ credentials });
          }
          return reply.redirect(options.endpoint).unstate(options.cookieName);
        }
        if (request.url.pathname === options.endpoint && request.method === 'get') {
          return reply(null, renderLogin(request, pluginOptions));
        }
        return reply.redirect(`${options.endpoint}?next=${request.url.path}`);
      }
    };
  });
};


lib/credentials.js
'use strict';
const _ = require('lodash');
const crypto = require('crypto');

// password-> credentials lookup table:
const passwordRepository = {};

// add credentials to the password object:
module.exports.registerCredentials = (password, salt) => {
  if (!password) {
    throw new Error('must set a password');
  }
  // if password is passed as a single string, there's only one password
  // (this will also trigger if the default options are used):
  if (typeof password === 'string' && typeof salt === 'string') {
    const pass = crypto.createHash('md5').update(password + salt).digest('hex');
    passwordRepository[password] = {
      encryptedPassword: pass
    };
  // otherwise assume that password is an object containing multiple password / credentials:
  } else {
    _.each(password, (credentials, passwordText) => {
      passwordRepository[passwordText] = {
        credentials,
        encryptedPassword: crypto.createHash('md5').update(passwordText + salt).digest('hex')
      };
    });
  }
};

// get credentials from the lookup object based on password:
module.exports.getCredentials = (password, useCookie) => {
  if (useCookie) {
    return _.find(passwordRepository, (o) => o.encryptedPassword === password);
  }
  return passwordRepository[password];
};


lib/registerRoutes.js
'use strict';
const renderLogin = require('./renderLogin.js');
const lookup = require('./credentials.js');

module.exports = (server, pluginOptions) => {
  // log in route:
  server.route({
    method: 'POST',
    path: pluginOptions.endpoint,
    config: {
      auth: false
    },
    handler: (request, reply) => {
      const credentials = lookup.getCredentials(request.payload.password);
      if (credentials) {
        return reply
          .redirect(request.payload.next || pluginOptions.successEndpoint)
          .state(pluginOptions.cookieName, credentials.encryptedPassword, {
            isSecure: pluginOptions.isSecure,
            ttl: pluginOptions.ttl,
            path: pluginOptions.path
          })
          .state(pluginOptions.cookieNameName, request.payload.name, {
            isSecure: pluginOptions.isSecure,
            ttl: pluginOptions.ttl,
            path: pluginOptions.path,
            strictHeader: false // allows 'name' to include whitespace and other special chars, etc)
          });
      }
      if (pluginOptions.logFailedAttempts) {
        server.log(['hapi-password', 'warning'], `Failed login at ${new Date}: Name: ${request.payload.name}`);
      }
      const nextString = request.payload.next ? `&next=${request.payload.next}` : '';
      reply.redirect(`${pluginOptions.endpoint}?error=1${nextString}`);
    }
  });
  // display login form route:
  if (typeof pluginOptions.loginForm === 'object') {
    server.route({
      method: 'GET',
      path: pluginOptions.endpoint,
      config: {
        auth: pluginOptions.strategyName,
        handler: (request, reply) => {
          if (request.auth.isAuthenticated) {
            return reply.redirect(pluginOptions.successEndpoint);
          }
          if (pluginOptions.logFailedAttempts) {
            server.log(['hapi-password'], `Failed login at ${new Date}: Name: ${request.payload.name}`);
          }
          reply(null, renderLogin(request, pluginOptions));
        }
      }
    });
  }
  // log out route:
  if (pluginOptions.logoutEndpoint) {
    server.route({
      path: pluginOptions.logoutEndpoint,
      method: 'GET',
      config: {
        auth: pluginOptions.strategyName
      },
      handler: (request, reply) => {
        const nextString = request.query.next ? encodeURI(request.query.next) : '/';
        reply
          .redirect(nextString)
          .unstate(pluginOptions.cookieName, { path: pluginOptions.path });
      }
    });
  }
};


lib/index.js
'use strict';
const Hoek = require('hoek');
const registerRoutes = require('./registerRoutes');
const registerScheme = require('./authScheme.js');

const schemeDefaults = {
  logFailedAttempts: false,
  schemeName: 'password',
  strategyName: 'password',
  queryKey: 'token',
  endpoint: '/login',
  logoutEndpoint: '/logout',
  isSecure: false,
  successEndpoint: '/',
  cookieName: 'hapi-password',
  cookieNameName: 'hapi-password-name',
  salt: 'aSalt',
  password: 'password',
  ttl: 1000 * 60 * 60 * 24 * 30, // 30 days
  loginForm: {
    name: 'Login',
    description: '',
    askName: false
  }
};

exports.register = (server, pluginOptions, next) => {
  pluginOptions = Hoek.applyToDefaults(schemeDefaults, pluginOptions);
  registerScheme(server, pluginOptions, schemeDefaults);
  server.auth.strategy(pluginOptions.schemeName, pluginOptions.strategyName, true, pluginOptions);
  registerRoutes(server, pluginOptions);
  next();
};

exports.register.attributes = {
  pkg: require('../package.json')
};
